### DISCUSSION TITLE
dvc repro: change output format?

### DISCUSSION THREAD

**Discussion by skshetry** [treeverse/dvc#5043] [New Features & Ideas]
Recently we made a change in the `dvc repro` output by removing the indent to `>` in front of the command scripts.

My reasoning for this is that when indenting the script, the output of the script starts from the leftmost part of the next line which makes it look ugly and congested.

Adding a ">" at least separates the DVC's output with the script's. 

#### Before:
```console
$ dvc repro
Running stage 'copy':
	echo "foo"
foo
	cp foo bar
	echo "bar"
bar
Generating lock file 'dvc.lock'
Updating lock file 'dvc.lock'

To track the changes with git, run:

	git add dvc.lock
Use `dvc push` to send your updates to remote storage.
```

Note that multiline commands weren't supported before.

#### After:
```console
$ dvc repro
Running stage 'copy':
> echo "foo"
foo
> cp foo bar
> echo "bar"
bar
Generating lock file 'dvc.lock'
Updating lock file 'dvc.lock'

To track the changes with git, run:

	git add dvc.lock
Use `dvc push` to send your updates to remote storage.
```

This brings me to the main point of this discussion.

1) Is it necessary to print the output during `repro`? Should that be hidden by default (and, show whenever an error occurs or via some flag)?
2) If we should print them by default, should we introduce some UI widget to separate the outputs?
   Like a pytest's divider/header or should we keep it to the minimum `>`/indent?

<details>
  <summary>Example 1</summary>
  
```
============== Running 'cp foo bar' ==============
foo
bar
```
</details>

<details>
  <summary>Example 2</summary>
  
JK :)
```python
   ___                  ___
  (o o)    Running     (o o)
 (  V  )    'copy'    (  V  )
/--m-m- -------------/--m-m- \
|                            |
|         echo "foo"         |
|         cp foo bar         |
|         echo "bar"         |
|____________________________|
foo
bar
```
</details>

@iterative/engineering

**Comment by jorgeorpinel** [comment]
OMG yes

```
   ___                  ___
  (o o)    Running     (o o)
 (  V  )    'copy'    (  V  )
/--m-m- -------------/--m-m- \
|                            |
|         echo "foo"         |
|         cp foo bar         |
|         echo "bar"         |
|____________________________|
```

**Reply by pared** [reply]
I would implement that for an easter egg, really, this is magnificent, @skshetry.

**Reply by jorgeorpinel** [reply]
Oh dear... This thing has threads 🙀

**Comment by jorgeorpinel** [comment]
> ```
> ...
> > cp foo bar
> > echo "bar"
> bar
> ...
> ```

Hmmm I guess `>` does look better than a tab when the command has output.

Unless the output could be padded with a tab too (but this way you need extra new lines):

```
...
	cp foo bar

	echo** "bar\nbaz"
	bar
	baz
...
```

UPDATE: So never mind. `>` is better than my suggestion. Hiding this.

**Comment by jorgeorpinel** [comment]
> Running stage 'copy':

(As discussed in https://github.com/iterative/dvc.org/pull/1980#discussion_r538027117) I think we should still say "with command(s)" though, in case users don;t realize that stages are comprised of terminal commands.

**Comment by jorgeorpinel** [comment]
And actually, on the `>` matter...

1. As @shcheklein mentioned somewhere, it makes it harder to copy/paste that line as a working command
2. A `>` character at the beginning of a command line is actually a valid command, which redirects nothing to a file name provided next, e.g.
  ```console
  $ > echo hi
  bash: hi: command not found
  $ ls
  echo
  ```

So I'm back to being of the opinion that some sort of indentation is better for this. My only idea is tabs (as we use in other commands) but each command output would need to be padded (with 2 tabs) also.

**Reply by pared** [reply]
So, maybe we should go with Example#1 or #2? When I am doing anything in DVC - (as a user) and running repro, I usually look for output of particular stage.  To do that I need to find the execution command, and marking it as in Examples makes it easily distiguishable.

**Reply by jorgeorpinel** [reply]
Yes as long as the commands are by themselves in a terminal line (easy to copy/paste) any format around them I don't see any problem with.

**Comment by pared** [comment]
> 1. Is it necessary to print the output during repro? Should that be hidden by default (and, show whenever an error occurs or via some flag)?

I would say we should not hide output by default. It would make sense in most cases, but in the most important one - long-lasting training, one usually needs to see the progress. But optional turn off for stages outputs sounds reasonable.
